miden-crypto/src/merkle/partial_mt/mod.rs

```diff
@@ -91,7 +91,43 @@ impl PartialMerkleTree {
     /// - If the depth is 0 or is greater than 64.
     /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
     /// - The provided entries contain an insufficient set of nodes.
+    /// - Any entry is an ancestor of another entry (creates hash ambiguity).
     pub fn with_leaves<R, I>(entries: R) -> Result<Self, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (NodeIndex, Word)> + ExactSizeIterator,
+    {
+        let entries: Vec<_> = entries.into_iter().collect();
+
+        // Validate that no entry is an ancestor of another entry.
+        // This prevents ambiguity: if we provide both a parent and its child,
+        // should we use the provided parent hash or compute it from the child?
+        //
+        // Algorithm: Put all provided nodes in a set, then for each node walk up toward
+        // the root checking if any ancestor is also in the set.
+        let provided_nodes: BTreeSet<NodeIndex> = entries.iter().map(|(node, _)| *node).collect();
+
+        for (node, _) in &entries {
+            let mut current = *node;
+            while !current.is_root() {
+                current = current.parent();
+                if provided_nodes.contains(&current) {
+                    return Err(MerkleError::EntryIsNotLeaf { node: current, descendant: *node });
+                }
+            }
+        }
```

Comment:
This works, but couldn't we just get the depth of the first node (from `NodeIndex), and then error out if depth of any of the subsequent nodes doesn't match this depth. This may make the code a bit more complicated, but should avoid the need for building the `BTreeSet` and the extra iteration over the entries.